blog/2020/12/package_latency.md

```diff
@@ -0,0 +1,718 @@
+                                                             firstmulticol)
+    else
+        outcols, finalcolnames = _combine_rows_with_first!(first, initialcols, 1, 1,
+                                                           f, gd, incols, targetcolnames,
+                                                           firstmulticol)
+    end
+    return idx, outcols, collect(Symbol, finalcolnames)
+end
+```
+
+Obviously, it helps if you know something about the internals of the package, which we, the authors of this blog post, do not.   But even if you are as naive as we are, you can pretty quickly discover a couple of key points.
+First, note that this method isn't entirely trivial; size alone is a (very rough) predictor of the cost of inference, and at nearly 50 lines this seems likely to be a method worth optimizing.  (Conversely, 1-line methods rarely require the kind of attention we're giving to `_combine_with_first`, unless there is extensive inlining.)  Second, there are several arguments that might be specialized on:
+
+- two arguments (`first` and `incols`) could potentially be `NamedTuple`s, and since `(x=1,)` and `(y=1,)` are different `NamedTuple` types, these arguments alone have potentially-huge possibility for specialization.  (If these are specialized for the particular column names in a DataFrame, then the scope for specialization is essentially limitless.) Indeed, a check `methodinstances(DataFrames._combine_with_first)` reveals that many of these specializations are for different `NamedTuple`s.
+
+- the `f::Base.Callable` argument is either a function or a type, again a potentially-limitless source of specialization. However, checking the output of `methodinstances`, you'll see that this argument is not specialized.  Presumably this is due to the major callers of `_combine_with_first` using a `@nospecialize` on their corresponding argument.  In this case, over-specialization does not seem to be a concern, but generally speaking function or type arguments are prime candidates for risk of over-specialization.
+
+Some strategies, like adding `@nospecialize`s, might be effective in reducing compile-time cost.  But without knowing a lot more about this package, it is difficult to know whether this might have undesirable effects on runtime performance.  So here we pursue a different strategy: let's focus on the fact that inference has to be performed for each unique combination of input types.  Since we have two highly-diverse argument types, the effect is essentially *multiplicative*.  But we also note that `incols` is just "passed through"; while we might want to preserve this type information, specializing on `incols` does not affect any portion of the body of this method other than the final calls to `_combine_tables_with_first!` or `_combine_rows_with_first!`.  Consequently, we may be wasting a lot of time specializing code that doesn't actually change dependening on the type of `incols`.
```

Comment:
Likewise here, depending on which style guide you prefer, "highly-diverse" -> "highly diverse", or not :).